Nullable reference types

jnm2 opened the pull request:
Closes #579.

I tried to keep the big initial annotation commit as straightforward as possible, saving more complex fixes for later commits. For files where the changes aren't absolutely obvious, please consider reviewing single commits at a time.

➡ Please comment if you'd like any of the types in these APIs to be annotated differently. I simply annotated them as nullable if `null` was tolerated in previous releases and left them non-nullable if `null` would have caused an exception. All nullable annotations (`?` and attributes) are a description of my understanding of the past rather than a statement from me about the future.

jnm2 commented on src/Shouldly.Tests/ConventionTests/ApprovePublicApi.ShouldlyApi.approved.cs:
```diff
@@ -7,156 +7,175 @@ public enum Case
     [System.ObsoleteAttribute("This class is only kept here for backwards compatibility. Please use ShouldAssert" +
         "Exception instead.")]
+    [System.Runtime.CompilerServices.NullableAttribute(0)]
     public class ChuckedAWobbly : System.Exception
```
@JosephWoodward This diff is going to look miserable until https://github.com/ApiApprover/ApiApprover/issues/54 is fixed. I don't know if that means this PR should wait?

JosephWoodward replied:
@jnm2 Shouldly is currently in a beta release of a new major version change so if we were going to introduce any breaking changes to the public API then now is a suitable time to do it (providing it didn't break backwards compatibility of course).

jnm2 replied:
Cool. I don't think any of these API changes are breaking changes. (Unless I unintentionally changed something besides NRT annotations.) If a project already has `<Nullable>enable</Nullable>` and upgrades to the annotated Shouldly, new warnings might show up now that the compiler knows for sure which inputs and outputs from Shouldly can and can't be null. I believe more of .NET Core is receiving annotations in 3.1, not a major version change.

jnm2 commented on src/Shouldly.Tests/Shouldly.Tests.csproj:
```diff
@@ -17,9 +15,8 @@
-    <PackageReference Include="PublicApiGenerator" Version="8.0.1" />
+    <PackageReference Include="PublicApiGenerator" Version="10.0.0-beta1" />
     <PackageReference Include="TestStack.ConventionTests" Version="3.0.1" />
-    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />
```
`dotnet-test` was discontinued: `https://github.com/xunit/xunit/issues/1835#issuecomment-431199286`

Version 2.3.1 has no rollforward policy and can only run on .NET Core 2.0 which is EOL.

jnm2 commented on src/Shouldly.Tests/Shouldly.Tests.csproj:
```diff
@@ -2,7 +2,6 @@
   <PropertyGroup>
     <TargetFrameworks>netcoreapp2.1</TargetFrameworks>
     <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);net472;</TargetFrameworks>
-    <LangVersion>latest</LangVersion>
```
Moved to Directory.Build.props so that it could be next to `<Nullable>enable</Nullable>`

jnm2 commented on src/Shouldly/Internals/EqualityComparer.cs:
```diff
@@ -33,10 +33,10 @@ public bool Equals(T x, T y)
+                    return object.Equals(y, null);
 
-                if (object.Equals(y, default(T)))
+                if (object.Equals(y, null))
```
`default(T)` is always `null` given the `if` condition. This is nicer than doing `default(T)!`.

jnm2 commented on src/Shouldly/Internals/EqualityComparer.cs:
```diff
@@ -33,10 +33,10 @@ public bool Equals(T x, T y)
             // Null?
             if (!type.IsValueType() || (type.IsGenericType() && type.GetGenericTypeDefinition().IsAssignableFrom(NullableType)))
```
This check would be simpler as `default(T)! is null`.

jnm2 commented on src/Shouldly/Internals/EqualityComparerAdapter.cs:
```diff
@@ -18,9 +18,9 @@ public EqualityComparerAdapter(IEqualityComparer<T> innerComparer)
+        public new bool Equals(object? x, object? y)
         {
-            return _innerComparer.Equals((T) x, (T) y);
+            return _innerComparer.Equals((T)x!, (T)y!);
```
The cast will always succeed when `T` is a nullable type at runtime (either a reference type or `Nullable<>`).

jnm2 commented on src/Shouldly/Internals/Is.cs:
```diff
@@ -3,18 +3,18 @@
     internal static class Is
     {
-        public static bool InRange<T>(T comparable, T @from, T to) where T : IComparable<T>
+        public static bool InRange<T>([DisallowNull] T comparable, [AllowNull] T @from, [AllowNull] T to) where T : IComparable<T>
```
`[AllowNull]` and `[DisallowNull]` are the only way to annotate these parameters since C# still does not allow `T?` here. (Allowing `T?` without a class or struct constraint would be a CLR change.)

jnm2 commented on src/Shouldly/ShouldlyExtensionMethods/ShouldBeEnumerableTestExtensions.cs:
```diff
@@ -102,62 +103,62 @@ public static void ShouldAllBe<T>(this IEnumerable<T> actual, [InstantHandle] Ex
         }
 
-        public static void ShouldBeEmpty<T>(this IEnumerable<T> actual)
+        public static void ShouldBeEmpty<T>([NotNull] this IEnumerable<T>? actual)
```
`[NotNull]` initially seems to conflict with the `?`, but there is no conflict. `[NotNull]` is indicating that if this call returns, the expression that had been passed to `actual` cannot have been null. At the same time, the method accepts null inputs without a compile-time warning.

You could also choose to go with `ShouldBeEmpty<T>(this IEnumerable<T> actual)` in which case the compiler would warn if the user does `foo.ShouldBeEmpty();` without first proving to the compiler that `foo` can't be null (or using `!`).

jnm2 commented on src/Shouldly/Internals/Numerics.cs:
```diff
@@ -14,7 +15,7 @@ internal static class Numerics
         /// <param name="obj">The object to check</param>
         /// <returns>true if the object is a numeric type</returns>
-        public static bool IsNumericType(Object obj)
+        public static bool IsNumericType([NotNullWhen(true)] object? obj)
```
`NotNullWhen` is used to alert the compiler's flow state analysis to this situation:

```cs
nullableThing = ...
if (IsNumericType(nullableThing))
{
    nullableThing.GetHashCode(); // No warning because IsNumericType returned true
}

nullableThing.GetHashCode(); // Compiler warning because `nullableThing` could be null at this point
```

jnm2 commented on src/Shouldly/Internals/ShouldThrowAssertionContext.cs:
```diff
@@ -4,18 +4,18 @@ namespace Shouldly
-            System.Diagnostics.StackTrace stackTrace = null,
-            [CallerMemberName] string shouldlyMethod = null) : base(shouldlyMethod, expected, actual, stackTrace)
+            System.Diagnostics.StackTrace? stackTrace = null,
+            [CallerMemberName] string shouldlyMethod = null!) : base(shouldlyMethod, expected, actual, stackTrace)
```
I don't think there's a situation where the default value `null` could be used.
https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callermembernameattribute#remarks

jnm2 commented on src/Shouldly/Internals/SourceCodeTextGetter.cs:
```diff
@@ -84,7 +84,7 @@ string GetCodePart()
                 var codeLines = string.Join("\n", File.ReadAllLines(FileName).Skip(LineNumber).ToArray());
 
-                var indexOf = codeLines.IndexOf(_shouldMethod);
+                var indexOf = codeLines.IndexOf(_shouldMethod!);
```
The way to avoid all these dammit operators (`!`) would be to pass the value between methods rather than making it a class field. The state of a mutable class field is also harder to reason about between methods for people as well as compilers, so I would recommend this.

jnm2 commented on src/Shouldly/Internals/StringHelpers.cs:
```diff
@@ -170,24 +170,24 @@ internal static string ToSafeString(this char c)
+        static string DelimitWith<T>(this IEnumerable<T> enumerable, string? separator) where T : class?
         {
-            return string.Join(separator, enumerable.Select(i => Equals(i, default(T)) ? null : i.ToString()).ToArray());
+            return string.Join(separator, enumerable.Select(i => Equals(i, null) ? null : i.ToString()).ToArray());
```
`default(T)` was always `null` already. Using `null` is even nicer now that the other option is `default(T)!`.

jnm2 commented on src/Shouldly/MessageGenerators/DictionaryShouldContainKeyAndValueMessageGenerator.cs:
```diff
@@ -14,6 +15,9 @@ public override bool CanProcess(IShouldlyAssertionContext context)
         public override string GenerateErrorMessage(IShouldlyAssertionContext context)
         {
+            Debug.Assert(context.Actual is IDictionary);
+            Debug.Assert(context.Key is object);
```
⚠ Not sure how you'd like to handle these. This method is relying on the fragile assumption that remote code will have set these fields, and these asserts silence the warnings this causes now that NRTs are enabled. The way I would handle it would be to require these values to be passed to the DictionaryShouldContainKeyAndValueMessageGenerator constructor.

jnm2 commented on src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs:
```diff
@@ -10,53 +11,72 @@ namespace Shouldly
 
         [ContractAnnotation("actual:null,expected:null => halt")]
-        public static void ShouldNotBe<T>(this T actual, T expected)
+        public static void ShouldNotBe<T>([AllowNull] this T actual, [AllowNull] T expected)
```
`[AllowNull]` is needed on each so that generic inference doesn't get upset when one expression is nullable and the other isn't.

jnm2 commented on src/Shouldly/ShouldlyExtensionMethods/ShouldBe/StringNullOrEmptyTestExtensions.cs:
```diff
@@ -1,43 +1,44 @@
 {
     public static partial class ShouldBeStringTestExtensions
     {
         [ContractAnnotation("actual:notnull => halt")]
```
Side note: I think these JetBrains annotations might actually be incorrect. If actual is `""`, a non-null value, the method does not halt.

jnm2 commented on src/Shouldly/ShouldlyExtensionMethods/ShouldBeBooleanExtensions.cs:
```diff
@@ -8,33 +9,33 @@ namespace Shouldly
     public static class ShouldBeBooleanExtensions
     {
-        public static void ShouldBeTrue(this bool actual)
+        public static void ShouldBeTrue([DoesNotReturnIf(false)] this bool actual)
```
`DoesNotReturnIf` is the annotation that powers scenario like this:

```cs
nullableThing.GetHashCode(); // Compiler warning, possible NRE
(nullableThing != null).ShouldBeTrue();
nullableThing.GetHashCode(); // No warning because ShouldBeTrue did not throw an exception
```

jnm2 commented on src/Shouldly/ShouldlyExtensionMethods/ShouldBeDictionaryTestExtensions.cs:
```diff
@@ -3,71 +3,83 @@
+    public static class ShouldBeDictionaryTestExtensions
     {
         public static void ShouldContainKey<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key)
+            where TKey : notnull
```
This is necessary in order to use `TKey` for `IDictionary<TKey, TValue>` due to the generic constraint added to `IDictionary<,>` in the BCL.

jnm2 commented on src/Shouldly/ShouldlyExtensionMethods/ShouldBeNullExtensions.cs:
```diff
@@ -9,38 +11,38 @@ namespace Shouldly
     {
         [ContractAnnotation("actual:notnull => halt")]
-        public static void ShouldBeNull<T>(this T actual)
+        public static void ShouldBeNull<T>([MaybeNull] this T actual)
```
If this method returns, the `actual` expression is definitely null. `[MaybeNull]` is the strongest the .NET annotations get. Threw it in for completeness.

```cs
var nonnullableThing = new object();
nonnullableThing.GetHashCode(); // No warning
nonnullableThing.ShouldBeNull();
nonnullableThing.GetHashCode(); // Warning that nonnullableThing may be null
```

jnm2 commented on src/Shouldly/ShouldlyMessage.cs:
```diff
@@ -134,7 +134,7 @@ internal class ExpectedOrderShouldlyMessage : ShouldlyMessage
 
     internal class ExpectedEquvalenceShouldlyMessage : ShouldlyMessage
```
This is a typo.

danielmarbach commented:
Beta 3 of PublicApiGenerator is on nuget

jnm2 commented:
@JosephWoodward Everything is a go as soon as you get a chance to review, and there's still the GitVersion failure on the Ubuntu in the master branch which is what the CI failure is about.

Is there anyone else I should ping for review? The more eyes the better for something like this, probably. :D

sungam3r commented on src/DocumentationExamples/DocumentationExamples.csproj:
```diff
@@ -17,7 +17,7 @@
     <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
     <PackageReference Include="xunit" Version="2.4.1" />
-    <PackageReference Include="PublicApiGenerator" Version="8.1.0" />
+    <PackageReference Include="PublicApiGenerator" Version="10.0.0-beta3" />
```
10.0.0-beta3 has a bug with `readonly` structs, it was fixed in beta4

jnm2 replied:
This doesn't affect Shouldly, but I added a commit updating to 10.0.0.

jnm2 commented:
@JosephWoodward Ready for review.

JosephWoodward commented:
@jnm2 Amazing work :D I'll try and go through this over the next few days. Great PR though!

jnm2 commented:
@JosephWoodward New force push. Two of the commits have already been merged in https://github.com/shouldly/shouldly/pull/596 and I fixed merge conflicts caused by https://github.com/shouldly/shouldly/pull/595.

JosephWoodward commented:
@jnm2 Nice! So is this good to go? (pending review of course)

jnm2 commented:
@JosephWoodward Yes!

jnm2 commented:
@JosephWoodward I merged with master to fix merge commits, then I recognized the build failure as being related to certain versions of Mono and Cake pre-0.37 and fixed that, then I fixed some new warnings that will crop up in VS 16.5.

I'm not sure reviewing and merging will get more appealing as time goes on, and I don't know what to do to make that easier.

jnm2 commented:
Pushing a merge commit to resolve conflicts soon.

jnm2 commented:
@SimonCropp I was going to push a merge commit to resolve the conflicts. Would it make more sense to wait till the dust settles from the PRs you're merging, or would there be another approach altogether that would work better for you?

SimonCropp commented:
i can hold off for a few hours. i i can handle the conflicts and merge for u

jnm2 commented:
@SimonCropp You might have better context than me for fixing the conflicts since you are familiar with what's happened on master in the meantime. That sounds best, but I'm also happy to do it. Once this PR merges, there is a good chance that most outstanding PRs will have merge conflicts too. The tradeoff is fixing things individually versus adding the equivalent work in bulk to this PR. Is one better than the other?

SimonCropp commented:
@jnm2 i manually merged here https://github.com/shouldly/shouldly/pull/634. there still are a few nullability warnings. i was wondering if you could have a look at the warnings on master, given you clearly know more about nullability  than me

jnm2 commented:
@SimonCropp For sure!
